alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportIndexMonitorAction.kt

```diff
@@ -492,16 +493,36 @@ class TransportIndexMonitorAction @Inject constructor(
                     )
                     return
                 }
-                request.monitor = request.monitor.copy(id = indexResponse.id)
-                var (metadata, created) = MonitorMetadataService.getOrCreateMetadata(request.monitor)
-                if (created == false) {
-                    log.warn("Metadata doc id:${metadata.id} exists, but it shouldn't!")
+                var metadata: MonitorMetadata?
+                try { // delete monitor if metadata creation fails, log the right error and re-throw the error to fail listener
+                    request.monitor = request.monitor.copy(id = indexResponse.id)
+                    var (monitorMetadata: MonitorMetadata, created: Boolean) = MonitorMetadataService.getOrCreateMetadata(request.monitor)
+                    if (created == false) {
+                        log.warn("Metadata doc id:${monitorMetadata.id} exists, but it shouldn't!")
+                    }
+                    metadata = monitorMetadata
+                } catch (t: Exception) {
+                    log.error("failed to create metadata for monitor ${indexResponse.id}. deleting monitor")
+                    client.execute(
+                        AlertingActions.DELETE_MONITOR_ACTION_TYPE,
+                        DeleteMonitorRequest(indexResponse.id, RefreshPolicy.IMMEDIATE)
+                    )
+                    throw t
                 }
-                if (request.monitor.monitorType == Monitor.MonitorType.DOC_LEVEL_MONITOR) {
-                    indexDocLevelMonitorQueries(request.monitor, indexResponse.id, metadata, request.refreshPolicy)
+                try {
+                    if (request.monitor.monitorType == Monitor.MonitorType.DOC_LEVEL_MONITOR) {
+                        indexDocLevelMonitorQueries(request.monitor, indexResponse.id, metadata, request.refreshPolicy)
+                    }
+                    // When inserting queries in queryIndex we could update sourceToQueryIndexMapping
+                    MonitorMetadataService.upsertMetadata(metadata, updating = true)
+                } catch (t: Exception) {
+                    log.error("failed to index doc level queries monitor ${indexResponse.id}. deleting monitor", t)
+                    client.execute(
+                        AlertingActions.DELETE_MONITOR_ACTION_TYPE,
+                        DeleteMonitorRequest(indexResponse.id, RefreshPolicy.IMMEDIATE)
+                    )
```

Comment:
Lets catch this as well and log that we couldnt do the monitor cleanup. Also we would need to delete the metadata that we created as well.

Reply:
delete monitor transport action deletes the metadata as well.